Refactor Base Schema to not require Donor/Specimen/Sample (#878)
overture-stack/SONG: docs/custom-schemas.md

# Custom Analysis Schemas

## Minimal Example

```json
{
 "name": "minimalExample",
 "options":{},
 "schema":{
     "type": "object",
     "required":[
         "experiment"
     ],
     "properties":{
        "experiment":{}
     }
  }
}
```

## Options

> The `options` property is not required, if provided extra validations for this analysis can be specified. Similarly, each property in `options` is also optional. If no value is provided for an option, a default configuration will be used for the analysis. If this is an update to an existing analysis type, you can omit any option and its value will be maintained from the previous version.

Reviewer: ```suggestion
The `options` property is not required. If provided, extra validations for this analysis can be specified. Similarly, each property in `options` is also optional. If no value is provided for an option, a default configuration will be used for the analysis. If this is an update to an existing analysis type, you can omit any option and its value will be maintained from the previous version.
```

```json
{
	"fileTypes":["bam", "cram"],
	"externalValidations":[
		{
			"url": "http://localhost:8099/",
			"jsonPath": "experiment.someId"
		}
	]
},
```

If you want to remove the previous value of an option so that this validation is no longer required, for instance removing the restriction on file types so that any file type could be provided, you should provide an empty list for that option.

> In the example below, both `fileTypes` and `externalValidations` properties are set to empty arrays, which means that these validations will not be applied to submitted analysis:

Reviewer: love it. very helpful description 👍

# Custom Analysis Schemas

## Minimal Example

```json
{
 "name": "minimalExample",
 "options":{},
 "schema":{
     "type": "object",
     "required":[
         "experiment"
     ],
     "properties":{
        "experiment":{}
     }
  }
}
```

## Options

> All properties in options are optional. If no value is provided, a default configuration will be used for the analysis. If this is an update to an existing analysis type, you can omit any option and the value will be maintained from the previous version.

Reviewer: ```suggestion
All properties in "options" are elective. If no value is provided, a default configuration (shown bellow) will be used for the analysis. If this is an update to an existing analysis type, you can omit any option and its value will be maintained from the previous version (if one is available).
```
is the following object a representation of the default options? added note here in case that is so, to make it explicit. otherwise, we could include a link to the file/code that contains said default options.

Reviewer: also, if I omit an option because I'm trying to clear it for whatever reason, but it was already set (and therefor the existing value will be maintained), should we explain here how to remove it?

```json
{
	"fileTypes":["bam", "cram"],
	"externalValidation":[
		{
			"url": "http://localhost:8099/",
			"jsonPath": "experiment.someId"
		}
	]
},
```

### File Types

`options.fileTypes` can be provided an array of strings. These represent the file types (file extensions) that are allowed to be uploaded for this type of analysis.

If an empty array is provided, then any file type will be allowed. If an array of file types is provided, then an analysis will be invalid if it contains files of a type not listed.

```json
{
	"options": {
		"fileTypes": ["bam","cram"]
	}
}
```

### External Validation

> External validations configure Song to check a value in the analysis against an external service by sending an HTTP GET request to a configurable URL. The URL needs to return 2XX status message to indicate that if the provided value "is valid", typically meaning that the value for this property is known by that service.

Reviewer: ```suggestion
External validations configure Song to check a value in the analysis against an external service, by sending an HTTP GET request to a configurable URL. The service should respond with a 2XX status message to indicate the value "is valid".
```

Author: Including all these changes in my next commit, some of them conflict with a change I made so I'll commit the suggestions but not through the GH UI.

As an example, if the project clinical data is being managed in a separate service, we can add an external validation to the donor id field of our custom scheme. This will send the donor id to the external service which can confirm that we have previously registered that donor.

This might look like the following:

```json
{
	"url": "http://example.com/{study}/donor/{value}",
	"jsonPath": "experiment.donorId"
}
```

The URL provided is a template, with two variables that will be replaced during validation. Song will replace the token `{value}` with the value read from the analysis at the property as defined in the `jsonPath`. Song will also replace the token `{study}` with the study ID for the Analysis.

Continuing the above example, if the following analysis was submitted:

```json
{
  "studyId": "ABC123",
  "analysisType": {
    "name": "minimalExample"
  },
  "files": [
    {
      "dataType": "text",
      "fileName": "file1.txt",
      "fileSize": 123,
      "fileType": "txt",
      "fileAccess": "open",
      "fileMd5sum": "595f44fec1e92a71d3e9e77456ba80d1"
    }
  ],
  "experiment": {
    "donorId": "id01"
  }
}
```

> Song would attempt to validate the donorId by sending a URL to `http://example.com/ABC123/donor/id01`.

Reviewer: ```suggestion
Song would attempt to validate the donorId by sending a validation request to `http://example.com/ABC123/donor/id01`.
```

> It is allowable to use either the `{study}` or `{value}` multiple times in the URL, each instance will be replaced by the corresponding value.

Reviewer: ```suggestion
The URL parsing allows using either the `{study}` or `{value}` placeholders multiple times (e.g. `http://example.com/{study}-{value}/{value}`), each instance will be interpolated accordingly.
```

> [!Warning]

> > The URL may cause errors in song if it contains any tokens matching the `{word}` format other than `{study}` and `{value}`

Reviewer: Changing this here for consistency with the rest of your additions... and now I realize I've been noticing we spell SONG/Song/song in different places (beyond this repo).
Is it worth making a pass in a different PR to unify those here? if so, which spelling should we use? in Overture docs, we've been using all caps. TBD

```suggestion
> The URL may cause errors in Song if it contains any tokens matching the `{word}` format other than `{study}` and `{value}`
```
